### ISSUE TITLE
Restore fails if feed is protected by user name and password

### ISSUE THREAD

**Issue by livarcocc** [NuGet/Home#5909]
_From @jmecosta on September 20, 2017 5:44_

# Cannot restore when a feed is protected

we have many feeds, some of them are protected. in non dotnet core applications, we can proceed if permissions are not set for the feed since it try to restores from the next configure feed. Now it just fails, 

D:\prod\proto\diffplex>dotnet restore
  Restoring packages for D:\prod\proto\diffplex\WebDiffer\WebDiffer.csproj...
  Restoring packages for D:\prod\proto\diffplex\Perf.DiffPlex\Perf.DiffPlex.csproj...
  Restoring packages for D:\prod\proto\diffplex\DiffPlex.ConsoleRunner\DiffPlex.ConsoleRunner.csproj...
  Restoring packages for D:\prod\proto\diffplex\Facts.DiffPlex\Facts.DiffPlex.csproj...
  Restoring packages for D:\prod\proto\diffplex\WebDiffer\WebDiffer.csproj...
  Restore completed in 23,02 ms for D:\prod\proto\diffplex\DiffPlex\DiffPlex.csproj.
  Retrying 'FindPackagesByIdAsyncCore' for source 'http://a-buildmgmt/httpAuth/app/nuget/v1/FeedService.svc/FindPackagesById()?id='System.ComponentModel.TypeConverter''.
  Response status code does not indicate success: 401 (Unauthorized).
  Retrying 'FindPackagesByIdAsyncCore' for source 'http://a-buildmgmt/httpAuth/app/nuget/v1/FeedService.svc/FindPackagesById()?id='System.ComponentModel.TypeConverter''.
  Response status code does not indicate success: 401 (Unauthorized).
C:\Program Files\dotnet\sdk\2.0.0\NuGet.targets(102,5): error : Failed to retrieve information about 'System.ComponentModel.TypeConverter' from remote source 'http://a-buildmgmt/httpAuth/app/nuget/v1/FeedService.svc/FindPackagesById()?id='System.ComponentModel.TypeConverter''.\r [D:\prod\proto\diffplex\DiffPlex.sln]
C:\Program Files\dotnet\sdk\2.0.0\NuGet.targets(102,5): error :   Response status code does not indicate success: 401 (Unauthorized). [D:\prod\proto\diffplex\DiffPlex.sln]

# General
D:\prod\proto\diffplex>dotnet --version
2.0.0

.NET Command Line Tools (2.0.0)

Product Information:
 Version:            2.0.0
 Commit SHA-1 hash:  cdcd1928c9

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.0.0\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0
  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d

thanks and regards

_Copied from original issue: dotnet/core#976_

**Comment by nkolev92** [comment]
@jmecosta 
There are multiple reasons your scenario does not work. 

1) Dotnet.exe restore does not have CredentialProvider support for package protected feeds. 
The only way dotnet.exe could authenticate right now is using nuget.config and clear text passwords. 

2) There is a different between how packages.config and package reference(net core) based projects treat unreachable sources. 

Due to the fact package reference based projects have to resolve package versions based on available versions on the feeds, it does not allow for sources to be unreachable.
By design as NuGet hard-fails when there is a failed source because this can potentially cause inconsistent restores based on which sources are available.

To work around this, you can use --ignore-failed-sources in dotnet.exe and NuGet.exe

**Comment by jmecosta** [comment]
@nkolev92 thanks for the clarification. but this means that we have always to use the --ignore-failed-sources for everything since it should be fairly common scenario that multiple feeds to be available to a project or projects (we use 5 different feeds, according to the internal provider). i notice also that if package is not found in feed it also fails. So basically the nominal behaviour of nuget in dotnet assumes only one nuget feed per project. That seems quite restrictive.

the inconsistencies will happen anyway, for example if package is present in both feeds and the feeds are configured by users in different order. This is something that cant be prevented by failing fast imo.

**Comment by nkolev92** [comment]
@jmecosta Not sure what you mean by 
```
So basically the nominal behaviour of nuget in dotnet assumes only one nuget feed per project. That seems quite restrictive.
```
NuGet does support multiple feeds per project, there are no restrictions there. You just need these feeds to be available otherwise you risk into 2 different restores resulting in 2 different behaviors. 

The ordering of sources in the config file does not impact which source is being used. 
NuGet will pick the most responsive one, the one from which it can download the package the fastest. 

There's a story https://github.com/NuGet/Home/issues/2572 that would eventually allow users to control the resolution per feed, but that's little ways down the road. 

There's also this issue which deals with adding dotnet.exe support for credential provider. https://github.com/NuGet/Home/issues/5447

**Comment by rrelyea** [comment]
Duplicate of https://github.com/NuGet/Home/issues/1851 - we continue to work with corefx team offline to get the needed cross platform support for this.

**Comment by jmecosta** [comment]
@rrelyea hum, i tried again and it only fails if a protected feed is enabled in nuget.config. For some reason earlier it was failing also if i had 3 feeds with the package only present in one of the feeds. But seems its working now in this case. So the story to support protected feeds should cover this case. thanks for the clarification

**Comment by netspark** [comment]
Is this fixed? Cant find in which release it is done

**Comment by nkolev92** [comment]
@netspark As indicated earlier, this was closed as  a duplicate of #1851.  
Please follow/subscribe to that issue for updates.
